mgm/src/gf/u64_soft.rs

```diff
@@ -0,0 +1,76 @@
+use core::num::Wrapping;
+use crate::Block;
+
+pub(crate) struct Element(u64, u64);
+
+impl Element {
+    pub(crate) fn new() -> Self {
+        Self(0, 0)
+    }
+
+    #[allow(clippy::cast_ptr_alignment)]
+    pub(crate) fn mul_sum(&mut self, a: &Block, b: &Block) {
+        let [a1, a0] = from_block(a);
+        let [b1, b0] = from_block(b);
+        let a2 = a1 ^ a0;
+        let b2 = b1 ^ b0;
+
+        let c = bmul64(a1, b1);
+        let d = bmul64(a0, b0);
+        let e = bmul64(a2, b2);
+        let t = c ^ d ^ e;
+        let v0 = d as u64;
+        let v1 = ((d >> 64) ^ t) as u64;
+        let v2 = (c ^ (t >> 64)) as u64;
+        let v3 = (c >> 64) as u64;
+
+        let d = v2 ^ (v3 >> 63) ^ (v3 >> 62) ^ (v3 >> 57);
+        self.1 ^= v0 ^ d ^ (d << 1) ^ (d << 2) ^ (d << 7);
+        self.0 ^= v1 ^ v3
+            ^ (v3 << 1) ^ (v3 << 2) ^ (v3 << 7)
+            ^ (d >> 63) ^ (d >> 62) ^ (d >> 57);
+    }
+
+    pub(crate) fn into_bytes(self) -> Block {
+        let mut block = Block::default();
+        block[..8].copy_from_slice(&self.0.to_be_bytes());
+        block[8..].copy_from_slice(&self.1.to_be_bytes());
+        block
+    }
+}
+
+/// Multiplication in GF(2)[X], truncated to the low 64-bits, with “holes”
+/// (sequences of zeroes) to avoid carry spilling.
+///
+/// When carries do occur, they wind up in a "hole" and are subsequently masked
+/// out of the result.
+fn bmul64(x: u64, y: u64) -> u128 {
```

Comment:
Note that I am using `u128` here instead of the reversing trick. It allows to simplify code a bit. Maybe it's worth to port this approacht to `polyval`?

Reply:
Sure.

(I'm also curious why all of this can't be implemented in terms of POLYVAL?)

Reply:
>I'm also curious why all of this can't be implemented in terms of POLYVAL?

It could be, but I am not sure if it will have the same performance as an "inlined" version. Plus there are some annoying differences between algorithms, e.g. MGM does not chain values like GHASH and Polyval and it uses a "direct" order of bits for mapping binary string to polynomial. I will look into unifying code bases later, but not now.

Reply:
Aah, ok!